app/models/spree/order_decorator.rb

```diff
@@ -1,20 +1,24 @@
 Spree::Order.class_eval do
+  # Update the invoice number before transitioning to complete
+  #
+  state_machine.before_transition :to => :complete, :do => :update_invoice_number!
```

Comment:
The transition [is wrapped in a transaction](https://github.com/pluginaweek/state_machine/blob/8a3ba81801782ac4ef9d4ace1209d8d50ffccdb0/lib/state_machine/integrations/active_record.rb#L104-L137), so when the rest of the transaction fails, the change would be rolled back if I understand correctly. That way, we don't randomly hit the database, but in a nice and clean transactional context, if I understand correctly. 

As for the Ruby 1.9 Syntax: No problem, I can fix that.